lib/fuser_mqubit.h

```diff
@@ -0,0 +1,1087 @@
+          break;
+        }
+      }
+
+      if (!all_prev_visited) {
+        prev_gates.resize(0);
+        return false;
+      }
+
+      if (cur_size + prev.qubits.size() > max_fused_size) continue;
+
+      if (all_prev_visited) {
+        scratch.push_back(std::move(prev));
+        prev_gates.push_back(&scratch.back());
+      }
+    }
+
+    return true;
+  }
+
+  static void GetAddedQubits(const GateF& fgate0, const GateF* fgate1,
+                             const GateF& fgate2, GateA& added) {
+    for (std::size_t i = 0; i < fgate2.qubits.size(); ++i) {
+      unsigned q2 = fgate2.qubits[i];
+
+      bool have_added_qubits = true;
+
+      for (auto q0 : fgate0.qubits) {
+        if (q0 == q2) {
+          have_added_qubits = false;
+          break;
+        }
+      }
+
+      if (have_added_qubits && fgate1 != nullptr) {
+        for (auto q1 : fgate1->qubits) {
+          if (q1 == q2) {
+            have_added_qubits = false;
+            break;
+          }
+        }
+      }
+
+      if (have_added_qubits) {
+        added.qubits.push_back(q2);
+        added.links.push_back(fgate2.links[i]);
+      }
+    }
+  }
+
+  // Fuse smaller gates with fgate back in gate time.
+  static void FusePrev(unsigned level, GateF& fgate) {
+    std::vector<const Gate*> gates;
+    gates.reserve(fgate.gates.capacity());
+
+    uint64_t bad_mask = 0;
+    auto links = fgate.links;
+
+    bool may_have_gates_to_fuse = true;
+
+    while (may_have_gates_to_fuse) {
+      may_have_gates_to_fuse = false;
+
+      std::sort(links.begin(), links.end(),
+                [](const Link* l, const Link* r) -> bool {
+                  auto lp = l->prev;
+                  auto rp = r->prev;
+
+                  if (lp != nullptr && rp != nullptr) {
+                    return lp->val->parent->time > rp->val->parent->time;
+                  } else {
+                    return lp != nullptr || rp == nullptr;
+                  }
+                });
+
+      for (auto link : links) {
+        if (link->prev == nullptr) continue;
+
+        auto p = link->prev->val;
+
+        if (!QubitsAreIn(fgate.mask, p->mask) || (p->mask & bad_mask) != 0
+            || p->visited > level || p->parent->unfusible) {
+          bad_mask |= p->mask;
+        } else {
+          AddGates(level, *p, gates);
+
+          may_have_gates_to_fuse = true;
+          break;
+        }
+      }
+    }
+
+    for (auto it = gates.rbegin(); it != gates.rend(); ++it) {
+      fgate.gates.push_back(*it);
+    }
+  }
+
+  // Fuse smaller gates with fgate forward in gate time.
+  static void FuseNext(unsigned level, GateF& fgate) {
+    uint64_t bad_mask = 0;
+    auto links = fgate.links;
+
+    bool may_have_gates_to_fuse = true;
+
+    while (may_have_gates_to_fuse) {
+      may_have_gates_to_fuse = false;
+
+      auto time = fgate.parent->time;
+
+      std::sort(links.begin(), links.end(),
+                [&time](const Link* l, const Link* r) -> bool {
```

Comment:
`time` is not used in the function body.

Reply:
Removed.